src/resmom/mom_comm.c

```diff
@@ -6156,14 +6525,459 @@ tm_request(int fd, int version)
+					/*
+					 ** Sending to another MOM.
+					 */
+					/* 
+					** Have we already added a stream for this mom?
+					** We don't want to open another stream to the same mom.
+					*/	
+					for (k = 0; k < momcnt; k++) {
+						if (tvnodeid == streamlist[k]) {
+							alreadyadded = 1;
+							break;
+						}
+					}
+
+					/* Open a stream to the mom if a stream isn't already open */
+					/* The host of the current vnode is phost, use phost */
+					if (phost->hn_stream == -1)
+						phost->hn_stream = tpp_open(phost->hn_host, phost->hn_port);
+					phost->hn_sister = SISTER_EOF;
+					if (phost->hn_stream < 0) {
+						sprintf(log_buffer, "tpp_open failed on %s:%d",
+							phost->hn_host, phost->hn_port);
+						log_joberr(-1, __func__, log_buffer, jobid);
+						continue;
+					}
+				
+					if (pbs_conf.pbs_use_mcast == 1 && !alreadyadded) {
+						/* add each of the tpp streams to the tpp mcast channel */
+						if (tpp_mcast_add_strm(mtfd, phost->hn_stream, FALSE) == -1) {
+							tpp_close(phost->hn_stream);
+							phost->hn_stream = -1;
+							sprintf(log_buffer, "mcast add node %d failed", tvnodeid);
+							log_joberr(-1, __func__, log_buffer, jobid);
+							continue;
+						}
+						/* add the mom to the list of moms who have streams added */
+						streamlist[momcnt] = tvnodeid;
+						momcnt++;
+					}
+					/* Add the IM_SPAWN_MULTI event to the list */	
+					if (nep == NULL) {
+						nep = event_alloc(pjob, IM_SPAWN_MULTI, fd, phost,
+							event, fromtask);
+						ep = nep;
+					} else {
+						ep = event_dup(nep, pjob, phost);
+					}
+					if (ep == NULL) {
+						sprintf(log_buffer, "failed to create event for %s",
+								phost->hn_host?phost->hn_host:"node");
+						log_err(errno, __func__, log_buffer);
+						tpp_close(phost->hn_stream);
+						phost->hn_stream = -1;
+						continue;
+					}	
+					phost->hn_sister = SISTER_OKAY;
+					num++;
+				} /* end of else case */
+			} /* end of looping through the nodes */
+
+			/* Save info off to the job struct */
+			tempinfo->ji_num_sisters = momcnt;
+			tempinfo->ji_taskid_list = temptaskid;
+			tempinfo->ji_nid_list = tempnid;
+			tempinfo->ji_taskid_index = index;
+			
+			if (num > 0) {
+				/* there's info to send to the sisters */
+				ret = im_compose(mtfd, pjob->ji_qs.ji_jobid,
+						get_jattr_str(pjob, JOB_ATR_Cookie),
+						IM_SPAWN_MULTI, ep->ee_event, fromtask,
+						found_empty_string ? IM_PROTOCOL_VER : IM_OLD_PROTOCOL_VER);
+				if (ret != DIS_SUCCESS) {
+					arrayfree(argv);
+					arrayfree(envp);
+					free(node_list);
+					free(tempinfo);
+					free(temptaskid);
+					free(tempnid);
+					free(streamlist);
+					tpp_mcast_close(mtfd);
+					goto done;
+				}
+			
+				/* send body of the message: 
+				** start with the identifier,
+				** then the number of items in node_list, 
+				** then send the list of where in a loop
+				** then argc, argv, env, etc.
+				*/
+				ret = diswui(mtfd, taskident);
+				if (ret != DIS_SUCCESS) {
```

Comment:
I think we can avoid the if (ret != DIS_SUCCESS) blocks being repeated. How about we check for ret being DIS_SUCCESS in every next step, like so:

ret = diswui(mtfd, taskident);
if (ret == DIS_SUCCESS) 
            ret = diswui(mtfd, list_size);
for (i = 0; i < list_size, ret == DIS_SUCCESS; i++)
           ret = diswui(mtfd, node_list[i]);
...